specs/agents/metrics-otel.md

```diff
@@ -0,0 +1,166 @@
+# OpenTelemetry Metrics API Support
+
+Agents SHOULD support collecting custom metrics via the OpenTelemetry Metrics API.
+This SHOULD be done by supplying a OpenTelemetry Metrics SDK compatible [MetricExporter](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#metricexporter). If a pull based approach is preferred, a [MetricReader](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#metricreader) can be used instead. This `MetricExporter/MetricReader` SHOULD transform the received metrics data to IntakeV2 metricsets and send them via the existing reporting mechanism.
+
+## Metric naming
+
+OpenTelemetry metric names are allowed to contain dots. This can lead to mapping problems in elasticsearch if for example metrics with the names `foo.bar` and `foo` are both ingested.
+Due to the nested object representation within metricsets, `foo` would need to be mapped as both an object and a number at the same time, which is not possible.
+
+To avoid such conflicts, agents MUST replace all dots in metric names with `_` upon export iff the `dedot_custom_metrics` configuration option is set to `true`:
```

Comment:
The current spec text here is supposed to act more as a starting point for a discussion.

The java agent already has the `dedot_custom_metrics` configuration, which is on by default. This currently only affects Micrometer metrics, as this is the only way of collecting custom metrics atm.

On the otherside, the APM server does **not** perform a dedoting of metric names when receiving metrics via OTLP.

This means with the current proposal, the transition from sending metrics via the OTLP exporter vs shipping them via the agent would not be seamless: because `dedot_custom_metrics` is on by default, the metric names would change.

There are some alternatives to consider:
 * Add a `dedot_custom_metrics` option to the APM-server and make it on by default (breaking change)
 * Change the default `dedot_custom_metrics` in the java agent to be disabled by default (breaking change)
 * Separate the `dedot_custom_metrics` in the java agent into two different config options, one for Otel and one for micrometer. Terrible idea imo, because I don't see a use case where these two would be configured differently.
 
 Please use this discussion to vote for what you think is best or to propose alternative solutions.

Reply:
Related: https://github.com/elastic/apm-dev/issues/856 (internal)

Reply:
After some more thinking and offline discussions, what about the following "optimistic" solution:
In long/mid term, we want to get rid of the dedoting in general by mapping metricsets in elasticsearch using `subobjects: false`. Until we get there, we should aim to minimize breaking-changes.

So for all agents which do not have the `dedot_custom_metrics ` config yet (=everyone but java):
 * Agents MAY add the `dedot_custom_metrics` configuration option, but don't need to (maybe added later ondemand if issues with users arise)
 * If `dedot_custom_metrics` is added, the default MUST be `false` -> no breaking change required later on when we support dots in names fully
 
 For all agents which already have the `dedot_custom_metrics` configuration (= the java agent):
  * Keep the configuration `dedot_custom_metrics` with its default as is and respect it for Otel metrics
  * As soon as we support dots in metric names fully, switch the default to `false` (breaking but mitigable change)
  
 Opinions?

Reply:
SGTM but proposing a slight variation that minimizes breaking changes in the future: Ignore the `dedot_custom_metrics` flag for OTel metrics in the Java agent and not implement it in other agents. As this can lead to data loss, we probably can't GA the feature (debatable as we did GA OTLP metrics intake). Work on support for `subobjects: false`: We could consider flattening the metric documents in APM Server, try getting https://github.com/elastic/elasticsearch/issues/88934 prioritized or contribute this feature to ES (👨‍🚀⏳?).

Then we only need to introduce the breaking change of changing the default for `dedot_custom_metrics` for Micrometer. I suppose that we don't really need the `dedot_custom_metrics` flag at that time anymore so we can deprecate it when we change the default to `false` and eventually remove it.

Reply:
> we probably can't GA the feature (debatable as we did GA OTLP metrics intake)

Due to the fact that we create separate metricsets for the Otel data instead of merging them with e.g. the agent collected jvm metricsets I think the mapping conflicts would not be too bad, at least not worse than for the OTLP intake. E.g. we would only loose Otel metrics data, all other metrics would continue to work normally. So I'm not sure whether this really is a GA blocker.